pkg/s3wrapper/client.go

```diff
@@ -161,6 +168,200 @@ func (c *S3Client) UploadFile(ctx context.Context, filePath, objectName string)
+	}
+	return offset, length, nil
+}
+
+func (c *S3Client) uploadPartCopy(log logrus.FieldLogger, partNum *int64, uploadID *string, sourceObjectKey string, destObjectKey string,
+	sourceStartBytes int64, sourceEndBytes int64) (*s3.CompletedPart, error) {
+	completedPartCopy, err := c.client.UploadPartCopy(&s3.UploadPartCopyInput{
+		Bucket:          aws.String(c.cfg.S3Bucket),
+		Key:             aws.String(destObjectKey),
+		CopySource:      aws.String(fmt.Sprintf("/%s/%s", c.cfg.S3Bucket, sourceObjectKey)),
+		CopySourceRange: aws.String(fmt.Sprintf("bytes=%d-%d", sourceStartBytes, sourceEndBytes)),
+		PartNumber:      partNum,
+		UploadId:        uploadID,
+	})
+	if err != nil {
+		err = errors.Wrapf(err, "Failed to copy part %d for file %s", partNum, destObjectKey)
+		log.Error(err)
+		return nil, err
+	}
+	return &s3.CompletedPart{ETag: completedPartCopy.CopyPartResult.ETag, PartNumber: partNum}, nil
+}
+
+func (c *S3Client) uploadIgnition(log logrus.FieldLogger, partNum *int64, uploadID *string, objectName,
+	ignitionConfig string, origContents []byte, areaLengthBytes int64) (*s3.CompletedPart, error) {
+	ignitionBytes := []byte(ignitionConfig)
+
+	// Create CPIO archive
+	archiveBuffer := new(bytes.Buffer)
+	cpioWriter := cpio.NewWriter(archiveBuffer)
+	if err := cpioWriter.WriteHeader(&cpio.Header{Name: "config.ign", Mode: 0o100_644, Size: int64(len(ignitionBytes))}); err != nil {
+		log.WithError(err).Errorf("Failed to write CPIO header")
+		return nil, err
+	}
+	if _, err := cpioWriter.Write(ignitionBytes); err != nil {
+		log.WithError(err).Errorf("Failed to write CPIO archive")
+		return nil, err
+	}
+	if err := cpioWriter.Close(); err != nil {
+		log.WithError(err).Errorf("Failed to close CPIO archive")
+		return nil, err
+	}
+
+	// Run gzip compression
+	compressedBuffer := new(bytes.Buffer)
+	gzipWriter := gzip.NewWriter(compressedBuffer)
+	if _, err := gzipWriter.Write(archiveBuffer.Bytes()); err != nil {
+		err = errors.Wrapf(err, "Failed to gzip ignition config")
+		log.Error(err)
+		return nil, err
+	}
+	if err := gzipWriter.Close(); err != nil {
+		err = errors.Wrapf(err, "Failed to gzip ignition config")
+		log.Error(err)
+		return nil, err
+	}
+
+	if int64(len(compressedBuffer.Bytes())) > areaLengthBytes {
+		err := errors.New(fmt.Sprintf("Ignition is too long to be embedded (%d > %d)", len(archiveBuffer.Bytes()), areaLengthBytes))
+		log.Error(err)
+		return nil, err
+	}
+
+	copy(origContents, compressedBuffer.Bytes())
+
+	contentLength := int64(len(origContents))
+	completedPartCopy, err := c.client.UploadPart(&s3.UploadPartInput{
+		Bucket:        aws.String(c.cfg.S3Bucket),
+		Key:           aws.String(objectName),
+		PartNumber:    partNum,
+		UploadId:      uploadID,
+		Body:          bytes.NewReader(origContents),
+		ContentLength: aws.Int64(contentLength),
+	})
+	if err != nil {
+		err = errors.Wrapf(err, "Failed to upload ignition for file %s", objectName)
```

Comment:
Maybe we should provide more metadata? `uploadID`, `S3Bucket`, `contentLength`..